sklearn/metrics/_safe_euclidean_sparse.pyx

```diff
@@ -0,0 +1,69 @@
+#cython: language_level=3
+#cython: boundscheck=False, cdivision=True, wraparound=False
+
+
+import numpy as np
+cimport numpy as np
+from cython cimport floating
+from libc.math cimport fmax
+
+
+np.import_array()
+
+
+ctypedef fused INT:
+    np.int32_t
+    np.int64_t
+
+
+def _euclidean_sparse_dense_exact(floating[::1] X_data,
+                                  INT[::1] X_indices,
+                                  INT[::1] X_indptr,
+                                  np.ndarray[floating, ndim=2, mode='c'] Y,
+                                  floating[::1] y_squared_norms):
+    cdef:
+        int n_samples_X = X_indptr.shape[0] - 1
+        int n_samples_Y = Y.shape[0]
+        int n_features = Y.shape[1]
+
+        int i, j
+
+        floating[:, ::1] D = np.empty((n_samples_X, n_samples_Y), Y.dtype)
+
+    for i in range(n_samples_X):
+        for j in range(n_samples_Y):
+            D[i, j] = _euclidean_sparse_dense_exact_1d(
+                &X_data[X_indptr[i]],
+                &X_indices[X_indptr[i]],
+                X_indptr[i + 1] - X_indptr[i],
+                &Y[j, 0],
+                y_squared_norms[j])
+
+    return np.asarray(D)
+
+
+cdef floating _euclidean_sparse_dense_exact_1d(floating *y_data,
+                                               INT *y_indices,
+                                               int y_nnz,
+                                               floating *x,
+                                               floating x_squared_norm) nogil:
+    """Euclidean distance between x dense and y sparse"""
+    cdef:
+        int i
+        floating xi
+        floating tmp = 0.0
+        floating result = 0.0
+        floating partial_x_squared_norm = 0.0
+    
+    # Split the loop to avoid unsafe compiler auto optimizations
+    for i in range(y_nnz):
+        xi = x[y_indices[i]]
+        partial_x_squared_norm += xi * xi
+
+    for i in range(y_nnz):
+        tmp = y_data[i] - x[y_indices[i]]
+        result += tmp * tmp 
```

Comment:
Would using float64 accumulators here help precision and reduce the (low) risk of overflow? Related to https://github.com/scikit-learn/scikit-learn/pull/13010

Reply:
Does it force conversion of `tmp` in `result += tmp*tmp` in that case ?
also we need to convert the result to float32 when we want float32.

I should test how it impacts performances.

Reply:
The cost is quite high. In this example:
```python
X = np.random.RandomState(0).random_sample((100000, 16)).astype(np.float32) 
Y = np.random.RandomState(1).random_sample((100, 16)).astype(np.float32) 
X[X<0.8] = 0 
X = csr_matrix(X) 
```
There's a 20% slowdown using a float64 accumulator.

I guess the precision is fine here since it's the exact calculation method.
I'm not sure about the risk of overflow. If the result is too big to be represented in float32, even if you can compute it's correct value on float64, when in the end you downcast it to float32 you'll get an inf anyway.